File: python/0076.py

class Solution:
    def minWindow(self, s, t):
        missing = len(t)
        need = dict()
        for c in t:
            need[c] = need.get(c, 0) + 1

        i = I = J = 0
        for j, c in enumerate(s, 1):
            if c not in need:
                continue
            if need[c] > 0:
                missing -= 1
            need[c] -= 1

            if missing > 0:
                continue

            while i < j:
                if s[i] in need:
                    if need[s[i]] >= 0:
                        break
                    else:
                        need[s[i]] += 1
                i += 1

            if not J or j - i < J - I:
                I, J = i, j

        return s[I:J]



print(Solution().minWindow("ADOBECODEBANC", "ABC"))
print(Solution().minWindow("a", "a"))
print(Solution().minWindow("a", "aa"))

File: cpp/0076.cpp
#include <string>
#include <iostream>
#include <vector>
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (t.size() > s.size()) return "";
        if (t.size() == 1) return (s.find(t[0]) == string::npos) ? "" : t;
        vector<int> need(128, 0);
        for (auto& c: t) need[c]++;
        int I = 0;
        int min_len = INT_MAX;
        int remain = t.size();
        for (int i = 0, j = 0; j < s.size(); j++) {
            remain -= need[s[j]] > 0 ? 1 : 0;
            need[s[j]]--;
            if (remain == 0) {
                while (i <= j) {
                    if (need[s[i]] >= 0) break;
                    need[s[i]]++;
                    i++;
                }
                int cur_len = j-i+1;
                if (min_len > cur_len) {
                    min_len = cur_len;
                    I = i;
                }
            }
        }
        return min_len == INT_MAX ? "" : s.substr(I, min_len);
    }
};


int main() {
    Solution sol;
    cout << sol.minWindow("ADOBECODEBANC", "ABC") << "\n";
    cout << sol.minWindow("a", "a") << "\n";
    cout << sol.minWindow("a", "aa") << "\n";
}
